execUtil.go

```
@@ -1,62 +1,155 @@
-  TaskIdCmdMap[task_id] = cmd
+	// validate task_id
+	if command, ok = taskIdCommandMap.Get(task_id); !ok {
+		s.Errors = []string{"invalid task_id"}
+		return s
+	}
 
-  log.Printf("RunCommand task_id=%s, pid=%d", task_id, TaskIdCmdMap[task_id].Process.Pid)
+	// TODO refactor this into un/finished process
+	s.Task_id = task_id
+	s.StartTime = new(time.Time)
+	*s.StartTime = command.StartTime
+	s.Finished = new(bool)
+	*s.Finished = command.GetFinished()
+	s.Errors = command.GetErrors()
 
-  return task_id
+	// cmd.Wait() has finished, append
+	if command.GetFinished() {
+		s.EndTime = new(time.Time)
+		*s.EndTime = command.EndTime
+		s.ExitCode = new(int)
+		*s.ExitCode = command.GetExitCode()
+	}
+
+	s.Output = command.StdoutBuff.Lines()
+
+	return s
 }
 
-// @param pid - stops process with this pid
-func StopProcess(task_id string) int {
+// RunCommand starts a process from command argument.
+func RunCommand(command string) StartResponse {
+	var s StartResponse
+
+	cmd := exec.Command(command)
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+
+	outBuf := NewOutput()
+	cmd.Stdout = io.MultiWriter(os.Stdout, outBuf)
+
+	err := cmd.Start()
+
+	if err != nil {
+		s.Error = err.Error()
+		return s
+	}
+
+	pgid, err := syscall.Getpgid(cmd.Process.Pid)
+
+	if err != nil {
+		s.Error = err.Error()
+		return s
+	}
+
+	task_id := hostname + "-" + strconv.Itoa(pgid) // TODO handle if Process or Pid nil
+	s.Task_id = task_id
+
+	taskIdCommandMap.Put(task_id, NewCommandWrapper(cmd, outBuf))
+
+	// async subroutine
+	go func() {
+		// TODO add append error to CommandWrapper, impl accessors and setters
+		err = cmd.Wait()
+		cw, _ := taskIdCommandMap.Get(task_id)
+		if err != nil {
+			cw.AppendError(err.Error())
+		}
+		cw.SetExitCode(cmd.ProcessState.ExitCode())
+		cw.SetFinished(true)
+	}()
+
+	return s
+}
+
+// StopProcess is called to end a previously started process.
+func StopProcess(task_id string) StopResponse {
```

Functions in this package take at least two responsibilities - the main responsibility of handling process start/query/stop loop and that of wrapping errors in some sort of response objects - apparently for the reason of being able to send them directly to client over the wire. This is sub-optimal - this package obviously has nothing to do with HTTP-based handling and should not hide errors inside a struct (at least errors that are not directly related to command), instead - return errors as expected and wrap them into responses/header codes in the package that implements HTTP layer.
This would mean returning error as usual:
```go
StopProcess(taskID string) (*StopResponse, error) {
```

---

yes, I was thinking along these lines but was trying to reduce the number of models. your suggestion makes sense.